src/pages/write-smart-contracts/billboard-tutorial.md

```diff
@@ -0,0 +1,281 @@
+alter the way the functions behave.
+
+Open the `tests/billboard_test.ts` file in your IDE. In this step, you will add a single automated test to exercise the
+`set-message` and `get-message` functions of the contract.
+
+```ts
+import { Clarinet, Tx, Chain, Account, types } from 'https://deno.land/x/clarinet@v0.10.0/index.ts';
+import { assertEquals } from 'https://deno.land/std@0.90.0/testing/asserts.ts';
+
+Clarinet.test({
+  name: 'Ensure that the message can be set',
+  async fn(chain: Chain, accounts: Map<string, Account>) {
+    let wallet_1 = accounts.get('wallet_1')!;
+
+    let assetMaps = chain.getAssetsMaps();
+    const balance = assetMaps.assets['STX'][wallet_1.address];
+
+    let block = chain.mineBlock([
+      Tx.contractCall('billboard', 'set-message', [types.utf8('testing')], wallet_1.address),
+      Tx.contractCall('billboard', 'get-message', [], wallet_1.address),
+      Tx.contractCall('billboard', 'set-message', [types.utf8('testing...')], wallet_1.address),
+      Tx.contractCall('billboard', 'get-message', [], wallet_1.address),
+    ]);
+
+    assertEquals(block.receipts.length, 4);
+    assertEquals(block.height, 2);
+
+    block.receipts[1].result.expectUtf8('testing');
+
+    block.receipts[3].result.expectUtf8('testing...');
+
+    assetMaps = chain.getAssetsMaps();
+    assertEquals(assetMaps.assets['STX'][wallet_1.address], balance - 210);
+  },
+});
+```
+
+Modify the default imports in the boilerplate to include `{ Clarinet, Tx, Chain, Account, types }` from the `clarinet`
+Deno library.
+
+Give the test a descriptive name using the `name` field, then modify the test function to include the `chain` and
+`accounts` arguments. Make sure that you define a type for each of the arguments to satisfy the Typescript compiler.
+
+Using the Clarinet library, define variables to get a wallet address principal from the Clarinet configuration, and the
+balance of that address on the chain.
+
+The functional part of the test is defined using the `chain.mineBlock()` function, which simulates the mining of a
+block. Within that function, the test makes 4 contract calls (`Tx.contractCall()`), 2 calls to `set-message` and 2 calls
+to `get-message`.
+
+Once the simulated block is mined, the test can make assertions about the chain state. This is accomplished using the
+`assertEquals()` function and the `expect` function. In this case, the test asserts that the once the simulated block
+is mined, the block height is now equal to `2`, and that the number of receipts (contract calls) in the block are
+exactly `4`.
+
+The test can then make assertions about the return values of the contract. The test checks that the result of the
+transaction calls to `get-message` match the string values that the calls to `set-message` contain. This covers the
+capability of both contract functions. Finally, the test asserts that STX were transferred from the transaction
+caller wallet, covering the price updating and token transfer.
```

Comment:
we should have the developer run the tests to see that they succeed (`clarinet test`)

Reply:
Also, my tests failed:

```
test Ensure that the message can be set ... FAILED (2ms)

failures:

Ensure that the message can be set
TypeError: Deno.core.jsonOpSync is not a function
    at fn (https://deno.land/x/clarinet@v0.10.0/index.ts:203:41)
    at asyncOpSanitizer (deno:runtime/js/40_testing.js:21:15)
    at resourceSanitizer (deno:runtime/js/40_testing.js:58:13)
    at exitSanitizer (deno:runtime/js/40_testing.js:85:15)
    at runTest (deno:runtime/js/40_testing.js:199:13)
    at Object.runTests (deno:runtime/js/40_testing.js:244:13)
    at file:///Users/alexandergraebe/Development/stacks/billboard-tutorial/clarinet/billboard-clarity/$deno$test.js:1:27

failures:

        Ensure that the message can be set

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out (21ms)
```

Reply:
This should be fixed if you pull the latest clarinet. The billboard was using an outdated clarinet typescript lib, thanks for reporting!

Reply:
I also added some assertions for demoing assertions on events:
```ts
        let [event] = block.receipts[0].events;
        let {sender, recipient, amount} = event.stx_transfer_event;
        sender.expectPrincipal("ST1J4G6RR643BCG8G8SR6M2D9Z9KXT2NJDRK3FBTK");
        recipient.expectPrincipal("ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.billboard");
        amount.expectInt(100);
```

https://github.com/hirosystems/clarinet/blob/9be1f72fd47d35d86209d797116c611d74ba33dc/examples/billboard/tests/billboard_test.ts#L28-L32